### ISSUE TITLE
GPU 0 always allocates memory when cuDNN RNN modules are used

### ISSUE THREAD

**Issue by jihunchoi** [pytorch/pytorch#2383]
In a multi-GPU environment, even when the ID of GPU (other than 0) is specified when calling `.cuda(..)`, it seems that cuDNN RNN modules (`nn.RNN`, `nn.GRU`) allocate some amount of memory in GPU 0.

The following is the code to reproduce the issue:
```
In [1]: import torch

In [2]: from torch import nn

In [3]: torch.__version__
Out[3]: '0.2.0_2'

In [4]: cudnn_lstm = nn.LSTM(3, 4)

In [5]: cudnn_lstm.cuda(2)
Out[5]: LSTM(3, 4)

In [6]: !nvidia-smi
Fri Aug 11 12:05:49 2017
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.59                 Driver Version: 384.59                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  TITAN Xp            On   | 00000000:02:00.0 Off |                  N/A |
| 23%   39C    P2    60W / 250W |    275MiB / 12189MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  TITAN Xp            On   | 00000000:04:00.0 Off |                  N/A |
| 23%   38C    P8    17W / 250W |     10MiB / 12189MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  TITAN Xp            On   | 00000000:83:00.0 Off |                  N/A |
| 23%   40C    P2    60W / 250W |    347MiB / 12189MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  TITAN Xp            On   | 00000000:84:00.0 Off |                  N/A |
| 23%   30C    P8     8W / 250W |     10MiB / 12189MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0     33124    C   /home/jhchoi/anaconda3/envs/ml/bin/python      265MiB |
|    2     33124    C   /home/jhchoi/anaconda3/envs/ml/bin/python      337MiB |
+-----------------------------------------------------------------------------+
```

It seems than modules other than cuDNN RNN, such as `nn.Linear` and `nn.Conv1d`, work as expected and do not consume memory on GPU 0.
Of course the amount of memory allocation is quite small and can be ignored in most cases, however it can be problematic when almost all memory of GPU 0 is already in use by some other processes.
(One can specify `CUDA_VISIBLE_DEVICES` to avoid this problem though)

**Comment by colesbury** [comment]
The memory usage is because CUDA is initialized on device 0 (the current device) even when the LSTM is moved to device 2. The cause is in `RNNBase.flatten_parameters()`:

https://github.com/pytorch/pytorch/blob/6648677acf1b653a2c80ee790dd85dc29664f6b0/torch/nn/modules/rnn.py#L78

`cudnn.get_handle()` gets the handle for the **current** device. We should probably switch the current device to the target (e.g. the device of `any_param`). This current behavior seems potentially buggy because we're using the cuDNN handle for device 0 to initialize descriptors used on device 2.

**Comment by jihunchoi** [comment]
I also found that training does not go well when `nn.DataParallel` wrapper is used with cuDNN RNN modules.
It prints warnings which say parameters are not flattened, and the aspect of loss changes of the multi-GPU case is not similar to that of the single-GPU case.
I think this problem is possibly related to the problem you stated.

**Comment by soumith** [comment]
fixed in master
